src/services/revocation.js

```diff
@@ -32,60 +32,183 @@ export class RevocationStatusServiceImpl {
       return error(new Failure('Revocation status check failed'))
     }
   }
 }
+
+/**
+ * Client-side proof chain verification utility
+ * Collects all CIDs in proof chain and checks them in parallel with cancellation
+ *
+ * @param {import('@ucanto/interface').Delegation[]} delegations - The delegations to verify
+ * @param {string} spaceDID - Space DID to validate delegation context
+ * @param {import('../types/env.js').Env} env - Environment configuration
+ * @param {number} concurrencyLimit - Max parallel requests (default: 5)
+ * @returns {Promise<{isValid: boolean, revokedDelegation?: string, reason?: string}>}
+ */
+async function hasValidDelegationChain (delegations, spaceDID, env, concurrencyLimit = 5) {
+  // Filter delegations that are for the expected space
+  const validDelegations = delegations.filter(delegation => {
+    return delegation.capabilities && delegation.capabilities.some(capability => {
+      return capability.with && capability.with === spaceDID
+    })
+  })
+
+  if (validDelegations.length === 0) {
+    return {
+      isValid: false,
+      reason: `No valid delegations found for space ${spaceDID}`
+    }
+  }
+
+  // Collect CIDs only from space-valid delegations (breadth-first)
+  /** @type {string[]} */
+  const cidsToCheck = []
+  /** @type {import('@ucanto/interface').Delegation[]} */
+  const queue = [...validDelegations] // Only start with space-valid delegations
+  const visited = new Set()
+
+  while (queue.length > 0) {
+    const current = queue.shift()
+    if (!current) continue
+    const cidStr = current.cid.toString()
+    if (visited.has(cidStr)) continue
+    visited.add(cidStr)
+    cidsToCheck.push(cidStr)
+
+    // Add proofs to queue for traversal, but only if they're for the same space
+    if (current.proofs) {
+      const spaceValidProofs = current.proofs.filter(proof => {
+        const p = /** @type {import('@ucanto/interface').Delegation} */ (proof)
+        return p.capabilities && p.capabilities.some(cap =>
+          cap.with && cap.with === spaceDID
+        )
+      })
+      queue.push(...spaceValidProofs.map(p => /** @type {import('@ucanto/interface').Delegation} */ (p)))
+    }
+  }
+
+  // Check all CIDs in parallel with concurrency limit and cancellation
+  const abortController = new AbortController()
+  let foundRevocation = null
+
+  /**
+   * Checks if a delegation is revoked
+   * @param {string} cid - The CID of the delegation to check
+   * @returns {Promise<{isValid: boolean, revokedDelegation?: string, reason?: string, revocationData?: any} | null>}
+   */
+  const checkCID = async (cid) => {
+    if (abortController.signal.aborted) return null
+    const response = await fetch(`${env.UPLOAD_SERVICE_URL}/revocations/${cid}`)
+    if (response.status === 200) {
+      // According to W3 revocations-check spec, 200 response means delegation is revoked
+      // We don't need to parse the CAR file, just trust the 200 status
```

Comment:
You do. You shouldn't trust the service - ideally the client library verifies the revocation. If you want to leave this as TODO then please add a comment and open the issue.

Reply:
What about this: https://github.com/storacha/w3infra/pull/507#discussion_r2276671788

Reply:
Got it. I will tackle the new lib implementation in another ticket: https://github.com/storacha/upload-service/issues/430